### ISSUE TITLE
Generated file site.webmanifest should be renamed to site.json for compatibility with IIS by default.

### ISSUE THREAD

**Issue by homestar9** [RealFaviconGenerator/realfavicongenerator#372]
In the generated favicon package there's a file called site.webmanifest. Unfortunately IIS can not serve files ending in .webmanifest by default. Since the file contains a JSON object I suggest renaming it to site.json.

**Comment by phbernard** [comment]
Oh :( That was the point of #342, part of https://github.com/RealFaviconGenerator/realfavicongenerator/milestone/5 

Could you tell me more about this issue? Which kind of error do you get when IIS is asked to serve such file?

**Comment by molant** [comment]
@homestar9 I don't think reverting this is a good idea. Even though Apache and IIS come with some configured mappings, depending on the version some could be outdated or not exist at all. That shouldn't be an excuse to not make the right thing and take the opportunity to teach something new to developers.

Adding new mime types is trivial:

* Apache:  `AddType application/manifest+json  webmanifest`
* IIS: `<mimeMap fileExtension="webmanifest" mimeType="application/manifest+json"/>`

Let's imagine for a second that you revert back to `.json`. Once IIS updates its mapping to use the right one this line will not be necessary, but there will be many sites that will be serving a `.json` file instead of the right `.webmanifest` one (and without [the benefits they could have](https://sonarwhal.com/docs/user-guide/rules/rule-manifest-file-extension/#why-is-this-important)). And what if the spec changes and starts enforcing the mimeType? There will be many sites that will not work because of this.

If you are worried about documentation in sonarwhal we have [server configurations for Apache and IIS](https://sonarwhal.com/docs/user-guide/rules/rule-content-type/#how-to-configure-the-server-to-pass-this-rule) that can be used.

There's a reason why there's a recommended mime type in the spec and I think that advice should be followed.

Also,  I'll chase down the right person in the IIS team to make sure the right mimeType is used with `.webmanifest`.

**Comment by homestar9** [comment]
Good point @molant. My argument was based on IIS not serving files that don't have a mapped MIME type so every user that wants to use a .webmanifest file would need to add a custom MIME type via IIS.  I remember back in the day when .mp4 wasn't a standard MIME type and had to be added manually. It took M$ years to add it.

@phbernard to answer your question, IIS will return a 404.3 error if you request a file that doesn't have a MIME mapping.

I briefly read the "benefits" link @molant provided and it sounds to me like the reason a new file extension (.webmanifest) was chosen was to allow for possible expansion or deviation from JSON standards in the future (please correct me if I'm wrong).  If I were in charge, I would keep it simple and use the .json file extension until the day comes that they decide to deviate from the spec.  That's just my opinion though.

You might be able to kill two birds with one stone by letting IIS users know that they will need to add a MIME type to IIS if they want that file to work.

**Comment by molant** [comment]
> If I were in charge, I would keep it simple and use the .json file extension until the day comes that they decide to deviate from the spec.

As someone that works on a browser I couldn't disagree more.
We all know that websites can take forever to update and that has a huge impact on browser vendors and how the web evolves. Some examples:
* [Firefox](https://hacks.mozilla.org/2016/09/firefox-49-fixes-sites-designed-with-webkit-in-mind-and-more/) and [Edge](https://msdn.microsoft.com/en-us/library/mt270097) had to add the `-webkit` prefix for some CSS properties because devs never updated their code once those become available. In most cases they were using tools that never got updated neither so even when the browsers supported it the tools was generating incomplete code.
* More recently, `Array.prototype.flatten` can't be added [because of a very old version of MooTools that is still in use](https://developers.google.com/web/updates/2018/03/smooshgate)

The spec recommends something, please let's do the right thing and follow the guidance.

> You might be able to kill two birds with one stone by letting IIS users know that they will need to add a MIME type to IIS if they want that file to work.

👍

**Comment by homestar9** [comment]
@molant I definitely respect your opinion as you obviously have more experience with W3C specs and browser implementations than I do.  My perspective comes from someone that authors web pages.  When I look at the spec, the part that seems the most important to me is the `<link rel="manifest" src="">`.   That tag indicates intent and if the browser expects a JSON object then a *.json file _sounds_ like the most reasonable file type to put in the `src` attribute.  However, I definitely yield my option to the experts in this matter as it sounds like there must be added value in having a new mime type and file extension.

**Comment by phbernard** [comment]
Thank you @homestar9 and @molant for your contributions.

The situation is actually different for IIS and Apache, because by default Apache will serve the manifest, although not with the suitable MIME type. But Chrome is smart enough to handle the situation.

@homestar9 I'm glad IIS returns a specific error message: at least you can quickly find out what's wrong. I suppose the message is very specific, like "Extension .webmanifest has no MIME mapping" ? What do you find when you google this exact message? If there is no specific answer (ie. answer for MIME mapping in general but not for the `.webmanifest` extension), I suggest you to post a question on Stackoverflow and answer it yourself. That will provide the minimum-hassle workaround to IIS users.

**Comment by homestar9** [comment]
@phbernard the IIS error message is as follows:
```The page you are requesting cannot be served because of the extension configuration. If the page is a script, add a handler. If the file should be downloaded, add a MIME map.```

Posting a Stackoverflow question/answer on the subject for other IIS users is a good idea.  Before I do so, what is the proper MIME type for a .webmanifest file? I couldn't find a definitive answer.  The W3C site references `application/manifest+json` but I saw a few different ones scattered throughout various blogs.

**Comment by molant** [comment]
[`application/manifest+json` is the right one](https://sonarwhal.com/docs/user-guide/rules/rule-content-type/#how-to-configure-the-server-to-pass-this-rule)

**Comment by homestar9** [comment]
Thank you.  I have posted the question and answer here:
https://stackoverflow.com/questions/49566446/how-can-i-have-iis-properly-serve-webmanifest-files-on-my-web-site
Please let me know if you would like me to make any revisions or if you see any errors.

**Comment by phbernard** [comment]
Superb! Thank you @homestar9. I couldn't help upvoting it ;)

Now wait a few days and let Google do its magic.

**Comment by brndto** [comment]
mmmmm
I've been renaming them to XML lately as I only generate favicons for android/iOS at the moment and both seem to be interpreting whatever's referenced in the manifest tag as JSON regardless of filetype or mimetype, in addition, IIS enables XML by default from memory since I haven't been having issues with it.

**Comment by benm-eras** [comment]
Would it be practical for some advice to IIS users to be added to the ASP.NET Core bit of the download page?

I eventually found that you can configure your application to service the `.webmanifest` file via the application [startup](https://stackoverflow.com/a/53266152/723645) file rather than messing about with IIS (or ISS Express for that matter which was what gave me my original problem).

**Comment by chrisdpratt** [comment]
The point of the `.webmanifest` extension is merely to give an easy target for sending the content type as `application/manifest+json`. The important part is not the extension, but rather the content-type. Even if you *did* use `.json` as the extension, you should set the content type as `application/manifest+json`, but that's going to be hard to do on just a one-off basis, hence the alternate extension. I personally feel this issue should be closed. The spec is clear about all of this and `.webmanifest` is actually preferred.

**Comment by JimWolff** [comment]
@chrisdpratt I agree, if this is what the specs say, it should definatly stay `.webmanifest` but "new extensions" usually cause IIS users "problems", because they have to be manually mapped, like if you want to use `.woff` files or when creating certificates with the `.well-known/acme-challenge` These things seems to become more and more common, so i think that pointing IIS users to the something like the answer @homestar9 gave should be enough.

It's unfortunate that it isn't plug-n-play but it really isn't hard to make those configurations for the user.
